File: EntropyBasedDiscretization.cpp
#include <iostream>
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <vector>
#include <algorithm>

using namespace std;
//计算一个数据集的entropy
double entropy( double S[][2],int from,int to,int class_types ){
    //只有一个元素，entropy=0
    if( from>=to )
        return 0.0;
    double ent=0.0;
    //计算每个类的概率
    int N=to-from+1;
    int* c_count=new int[class_types];
    memset(c_count,0,class_types*sizeof(int));
    for( int i=from;i<=to;i++ ){
        c_count[(int)S[i][1]]+=1;
    }
    double pro_i=0.0;
    //计算entropy, 以2为底
    for( int i=0;i<class_types;i++ ){
        if( c_count[i]>0 ){
            pro_i=(double)c_count[i]/N;
            ent-=(pro_i*log2(pro_i));
        }
    }
    delete[] c_count;
    return ent;
}
//计算一个划分的entropy
double entropyWithT( double S[][2],int from,int to,int T,int class_types){
    double ent1=entropy(S,from,T,class_types);
    double ent2=entropy(S,T+1,to,class_types);
    int N=to-from+1;
    double entST=(double)(T-from+1)/N*ent1+(double)(to-(T+1)+1)/N*ent2;
    //如果T==to，返回的是entropy(S)
    return entST;
}
//计算二分划分的最佳划分点
int bestCutPoint( double S[][2],int from,int to,int class_types ){
    int T=-1;double min_entST;double entST;
    for( int i=from;i<=to;i++ ){
        //找到相同值的右边界
        while( i<to && S[i]==S[i+1] )
            i++;
        entST=entropyWithT(S,from,to,i,class_types);
        //minimize Entropy(S,T), means maximize information gain
        if( T==-1 ){
            T=i;min_entST=entST;
        }else if( entST<min_entST ){
            T=i;min_entST=entST;
        }
    }
    return T;
}
//计算不同的类标数目
int uniqueClass( double S[][2],int from,int to,int class_types ){
    if( from==to )
        return 1;
    int* c_count=new int[class_types];
    memset(c_count,0,class_types*sizeof(int));
    for( int i=from;i<=to;i++ ){
        c_count[(int)S[i][1]]+=1;
    }
    int uniqueClass=0;
    for( int i=0;i<class_types;i++ ){
        if( c_count[i]!=0 )
            uniqueClass+=1;
    }
    return uniqueClass;
}
//Recursive Minimal Entropy Partioning
void partition(double S[][2],int from,int to,vector<int> &Ts,int class_types){
    //只有一个元素，没有划分的必要
    if( from>=to )
        return ;
    int N=to-from+1;
    int T=bestCutPoint(S,from,to,class_types);
    if( T==to )//最佳划分点是末尾元素，相当于没有划分
        return ;
    //检查Minimal Description Length Principle
    double gainST=entropy(S,from,to,class_types)-entropyWithT(S,from,to,T,class_types);
    //统计每部分中类标的数目,和entropy
    int k=uniqueClass(S,from,to,class_types);
    int k1=uniqueClass(S,from,T,class_types);
    int k2=uniqueClass(S,T+1,to,class_types);
    double entS=entropy(S,from,to,class_types);
    double entS1=entropy(S,from,T,class_types);
    double entS2=entropy(S,T+1,to,class_types);
    //计算下界
    double delta=log2(pow(3,k)-2)-(k*entS-k1*entS1-k2*entS2);
    double threshold=log2(N-1)/N+delta/N;
    if( gainST<= threshold ){
        return ;
    }else{
        //分割点记录
        Ts.push_back(T);
        //递归处理左右两部分
        partition(S,from,T,Ts,class_types);
        partition(S,T+1,to,Ts,class_types);
    }
}
//基于Entropy和Minimal Description Length的离散化方法
int discretization(double S[][2],int from,int to,int class_types){
    vector<int> Ts;
    //调用划分
    partition(S,from,to,Ts,class_types);
    //划分点排序
    sort(Ts.begin(),Ts.end());
    cout<< "-----划分点有" << Ts.size() << "个" <<endl;
    for( int i=0;i<Ts.size();i++ ){
        cout<< Ts[i] <<endl;
    }
    cout<< "-----离散化结果："<<endl;
    for( int i=0;i<Ts.size();i++ ){
        int po_left=(i==0)?from:Ts[i-1]+1;
        int po_right=Ts[i];
        for( int j=po_left;j<=po_right;j++ ){
            //离散化结果
            cout<< i <<endl;
        }
    }
    for( int i=Ts[Ts.size()-1]+1;i<=to;i++ ){
        //离散化结果
        cout<< Ts.size() <<endl;
    }
}
int main()
{
    //double S[7][2]={{1,0},{2,0},{3,1},{4,1},{5,2},{6,2},{7,2}};
    //double S[7][2]={{1,0},{2,0},{3,1},{4,1},{5,2},{6,2},{7,2}};
    double S[15][2]={{1,0},{2,0},{3,0},{4,0},{5,1},{6,1},{7,1},{8,1},{9,1},{10,2},{11,2},{12,2},{13,1},{14,2},{15,2}};
    //double entS=entropy(S,0,6,2);
    //cout<< entS <<endl;
    //cout<< infoGain(S,0,6,2,2,entS) <<endl;
    //cout<< bestCutPoint(S,0,6,2) <<endl;

    //vector<int> Ts;
    //partition(S,0,6,Ts,3);
    //cout<< "-----划分点有" << Ts.size() << "个" <<endl;
    //for( int i=0;i<Ts.size();i++ ){
    //    cout<< Ts[i] <<endl;
    //}

    discretization(S,0,14,3);
    return 0;
}

File: entropy_based_discretization.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Time    : 2018/5/3 8:53
# @Author  : WangliLin
# @Site    : 
# @File    : entropy_based_discretization.py
# @Software: PyCharm

from __future__ import division
from math import log
import numpy as np

'''
计算一个数据集的entropy
'''
def entropy(S, start, end, class_types):
    if (start > end):
        return 0.0
    ent = 0.0
    N = end - start + 1
    c_count = {}
    for i in range(class_types):
        c_count[i] = 0
    for i in range(start, end+1):
        c_count[S[i][1]] += 1
    pro_i = 0.0
    for i in range(class_types):
        if c_count[i] > 0:
            pro_i = c_count[i]/N
            ent_i = -(pro_i * log(pro_i, 2))
            ent += ent_i
            # print("class{0} count: {1}/{2} pro: {3} ent: {4}".format(i, c_count[i], N, pro_i, ent_i))
    del c_count
    # print("ent of this dataset:{}".format(ent))
    return ent

'''
计算一个划分的entopy
'''
def entropy_with_T(S, start, end, T, class_types):
    ent1 = entropy(S, start, T, class_types)
    ent2 = entropy(S, T+1, end, class_types)
    N = end - start + 1;
    entST = (T - start + 1) / N * ent1 + (end - (T+1) + 1) / N * ent2;
    print("entST with {0} of this dateset: {1}".format(T, entST))
    return entST

'''
计算二分划分的最佳划分点
'''
def best_cut_point(S, start, end, class_types):
    T = -1
    min_entST = 0
    entST = 0
    for i in range(start, end):
        print("i: {}".format(i))
        while (i < end) & (S[i] == S[i+1]):
            i += 1
        entST = entropy_with_T(S, start, end, i, class_types)
        if T == -1:
            T = i
            min_entST = entST
        elif entST < min_entST:
            T = i
            min_entST = entST
    return T

'''
计算不同的类别数目
'''
def unique_class(S, start, end, class_types):
    if start == end:
        return 1
    c_count = {}
    for i in range(class_types):
        c_count[i] = 0
    for i in range(start, end+1):
        c_count[S[i][1]] += 1
    unique_class = 0
    for i in range(class_types):
        if c_count[i] != 0:
            unique_class += 1
    return unique_class

'''
Recursive Minimal Entropy Patitioning
'''
def partition(S, start, end, Ts, class_types):
    if start >= end:
        return Ts
    N = end - start + 1
    T = best_cut_point(S, start, end, class_types)
    if T == end:
        return Ts
    gainST = entropy(S, start, end, class_types) - entropy_with_T(S, start, end, T, class_types)
    # 统计每部分中类别的数目 + entropy
    k = unique_class(S, start, end, class_types)
    k1 = unique_class(S, start, T, class_types)
    k2 = unique_class(S, T+1, end, class_types)
    entS = entropy(S, start, end, class_types)
    entS1 = entropy(S, start, T, class_types)
    entS2 = entropy(S, T+1, end, class_types)
    # print("----------------------")
    # print(N)
    # print(log(pow(3, k) - 2, 2))
    # print(k * entS - k1 * entS1 - k2 * entS2)
    delta = log(pow(3, k) - 2, 2) - (k * entS - k1 * entS1 - k2 * entS2)
    print("delta: {}".format(delta))
    threshold = (log(N-1, 2) + delta) / N
    # if gainST <= 0:
    print("-----------wwwww--------")
    print("threshold: {0} gainST: {1}".format(threshold, gainST))
    if gainST <= threshold:
    # if len(Ts) >= 2:
        return Ts
    else:
        Ts.append(T)
        print("----------")
        print(Ts)
        partition(S, start, T, Ts, class_types)
        partition(S, T+1, end, Ts, class_types)

def discretization(S, start, end, class_types):
    Ts = []
    partition(S, start, end, Ts, class_types)
    print("-----www-----")
    print(Ts)
    print("划分点为： {}".format(Ts))

def df2list(df, feature, label):
    df = df.sort_values(by=feature)
    list_all = []
    for i in range(df.shape[0]):
        # loc: 具体索引值 iloc： 根据其标签所在位置
        list_all.append([df.iloc[i][feature], df.iloc[i][label]])
    return list_all

if __name__ =='__main__':
    S = [[7,1], [5,1], [3,0], [4,0], [1,1], [2,0], [3,1], [0,1]]
    # ent = entropy(S, 0, 7, 2)
    # print("ent of this dataset:{}".format(ent))
    # entST = entropy_with_T(S, 0, 7, 3, 2)
    # print(entST)
    T = best_cut_point(S, 0, 7, 2)
    print(T)
    discretization(S, 0, 7, 2)